### PR TITLE
Add ADR describing an external API layer

### PR DESCRIPTION
**Pull request by rahearn**
## Description of change

* Added ADR describing pros and cons of servicing external API consumers via a new set of API endpoints
* Added OpenAPI documentation for the first external API endpoint
* Added OpenAPI documentation for cookie and bearer authentication methods

## How to test

N/A

## Issue(s)

* https://github.com/OHS-Hosting-Infrastructure/complaint-tracker/issues/147

### Every PR

- [x] Meets issue criteria
~- [ ] JIRA ticket status updated~
~- [ ] Code is meaningfully tested~
~- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)~
- [x] API Documentation updated
~- [ ] Boundary diagram updated~
~- [ ] Logical Data Model updated~
- [x] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: docs/adr/0015-external-facing-api-design-and-implementation.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Pros:
+
+* already implemented
+* well documented
+
+Cons:
+
+* API is highly coupled to TTA Hub UI needs, will become more difficult to change rapidly
+* authentication is based on sessions stored in browser cookies
+
```
Comment (jasalisbury) [inline docs/adr/0015-external-facing-api-design-and-implementation.md:31]: Another con would be authorization is also based on the user list contained in the TTA hub, and managed by hub admins. Seems like a fair to assume not all users of the external API are going to be users in the TTA hub.

File: docs/openapi/paths/api/schemas.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+          type: string
+          enum: ["activityReports"]
+        attributes:
+          type: object
+          properties:
+            author:
+              description: The creator of this activity report
+              $ref: "#/components/schemas/userObject"
+            collaborators:
+              description: Collaborating user ids
```
Comment (jasalisbury) [inline docs/openapi/paths/api/schemas.yaml:33]: Looks like the array here will contain the user id and display name

File: docs/openapi/paths/api/schemas.yaml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+              type: integer
+              minimum: 0
+              maximum: 13
+            reportCreationDate:
+              type: string
+              format: date-time
+            reportLastSaved:
+              type: string
+              format: date-time
+            # is there an important distinction between last updated vs last saved?
```
Comment (jasalisbury) [inline docs/openapi/paths/api/schemas.yaml:65]: Last saved is a date with the format `MM/DD/YYYY` (in UTC) and updated at is a full datetime (also in UTC)

File: docs/openapi/paths/api/schemas.yaml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+              description: How long the activity took to complete, in hours
+            endDate:
+              type: string
+              format: date
+            reason:
+              type: array
+              description: Why the activity was conducted
+              items:
+                type: string
+            recipient:
```
Comment (jasalisbury) [inline docs/openapi/paths/api/schemas.yaml:53]: Is this field the grant names that were the recipient of the TTA?

**Review by jasalisbury — COMMENTED** [review]
Looks good!

One thing I'm wonder about is authorization. For this specific API I think it is fine for all users that have an HSES login to have access. However I could see APIs for other systems, or even the hub, wanting to restrict access to some APIs. The example I keep thinking off (which may not be very realistic) is if there is an external API exposing monitoring data we may not want every HSES user to have access to it. Like I said doesn't really apply to this API!

File: docs/openapi/paths/api/schemas.yaml (hunk 3 above)
Comment (rahearn) [reply docs/openapi/paths/api/schemas.yaml:65]: OK, so we can probably drop last saved and calculate that ourselves from the updated at?

File: docs/openapi/paths/api/schemas.yaml (hunk 4 above)
Comment (rahearn) [reply docs/openapi/paths/api/schemas.yaml:53]: Yes, but I'm going to drop that field for right now. We don't need it right away and it's more properly situated in the `relationships` key anyway. Easier to add things later than move/change things.

File: docs/openapi/paths/api/schemas.yaml (hunk 3 above)
Comment (jasalisbury) [reply docs/openapi/paths/api/schemas.yaml:65]: Yeah you'd need the time portion if you wanted to display the date in local time (01-01-2000T23:00:00 CST is 01-02-2000 UTC)

**Comment by rahearn** [comment]
> One thing I'm wonder about is authorization. For this specific API I think it is fine for all users that have an HSES login to have access. However I could see APIs for other systems, or even the hub, wanting to restrict access to some APIs.

I think we can address in the short term with utilizing the existing TTA Hub roles, so a user would have to create a TTA Hub account to see details. In the longer term, I'd like to see centralizing those roles more fully in HSES, or breaking out user login to be a standalone service that TTA Hub, Complaint Tracker, and HSES _all_ use.

**Comment by kryswisnaskas** [comment]
Looks good and thanks for starting with an ADR and an OpenApi doc. 👍 
A couple of comments - Authorization is also on my mind. We need to ensure that the APIs are only available to users having the right permissions to view. Even though we are only providing a minimal set of data now, we still don't want to make it available to all HSES users. I think the plan with having the users create TTA Hub accounts is going to work.

> I think we can address in the short term with utilizing the existing TTA Hub roles, so a user would have to create a TTA Hub account to see details.
👍 

I'd like to see centralizing those roles more fully in HSES, or breaking out user login to be a standalone service that TTA Hub, Complaint Tracker, and HSES all use.
Interesting idea.

**Review by kryswisnaskas — APPROVED** [review]

**Pull request merged**